contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,322 @@
+  function _addTrustedProvisioner(address _token) internal {
+    trustedProvisioners.add(_token);
+    emit TrustedProvisionerAdded(_token);
+  }
+
+  function removeTrustedProvisioner(address _token) external onlyOwner {
+    trustedProvisioners.remove(_token);
+    emit TrustedProvisionerRemoved(_token);
+  }
+
+  function provisionPrepaidCard(address customer, bytes32 sku)
+    external
+    nonReentrant
+    returns (bool)
+  {
+    require(!paused, "Contract is paused");
+    require(
+      trustedProvisioners.contains(msg.sender),
+      "Only trusted provisioners allowed"
+    );
+    require(asks[sku] > 0, "Can't provision SKU with 0 askPrice");
+
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    string memory customizationDID = skus[sku].customizationDID;
+    uint256 priceToCreatePrepaidCard = prepaidCardManager.priceForFaceValue(
+      token,
+      faceValue
+    ); // in Wei
+
+    address issuer = skus[sku].issuer;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    require(
+      balance[issuerSafe][token] >= priceToCreatePrepaidCard,
+      "Not enough balance"
+    );
+
+    uint256[] memory issuingTokenAmounts = new uint256[](1);
+    uint256[] memory spendAmounts = new uint256[](1);
+
+    issuingTokenAmounts[0] = priceToCreatePrepaidCard;
+    spendAmounts[0] = faceValue;
+
+    balance[issuerSafe][token] -= priceToCreatePrepaidCard;
+
+    IERC677(token).safeTransferAndCall(
+      prepaidCardManagerAddress,
+      priceToCreatePrepaidCard,
+      abi.encode(
+        customer,
+        issuingTokenAmounts,
+        spendAmounts,
+        customizationDID,
+        address(0), // marketAddress - we don't need it in this case
+        issuer,
+        issuerSafe
+      )
+    );
+
+    emit PrepaidCardProvisioned(customer, sku);
+
+    return true;
+  }
+
+  function getQuantity(bytes32 sku) public view returns (uint256) {
+    PrepaidCardManager prepaidCardManager = PrepaidCardManager(
+      prepaidCardManagerAddress
+    );
+
+    address token = skus[sku].issuingToken;
+    uint256 faceValue = skus[sku].faceValue;
+    address issuerSafe = skus[sku].issuerSafe;
+
+    uint256 price = prepaidCardManager.priceForFaceValue(token, faceValue);
+
+    return balance[issuerSafe][token] / price;
+  }
+
+  function setAsk(
+    address issuer,
+    bytes32 sku,
+    uint256 askPrice
+  ) external returns (bool) {
+    require(skus[sku].issuer != address(0), "Non-existent SKU");
+    require(skus[sku].issuer == issuer, "SKU not owned by issuer");
+    require(msg.sender == skus[sku].issuerSafe, "Only issuer safe can set ask");
+
+    asks[sku] = askPrice;
+
+    emit AskSet(issuer, skus[sku].issuingToken, sku, askPrice);
+    return true;
+  }
+
+  function addSKU(
+    uint256 faceValue,
+    string memory customizationDID,
+    address token
+  ) external returns (bool) {
```

Comment:
this is also only called by prepaid cards not generic safes You'll need an `onlyHandlers` modifer here like we do in the V1 version of this https://github.com/cardstack/card-pay-protocol/blob/main/contracts/PrepaidCardMarket.sol#L192